addons/common/fnc_uniqueUnitItems.sqf

```diff
@@ -22,28 +23,66 @@ Returns:
     Array of item classnames <ARRAY>
 
 Author:
-    Dedmen
+    Dedmen, commy2
 ---------------------------------------------------------------------------- */
-SCRIPT(uniqueUnitItems);
 
-params [["_unit", objNull, [objNull]], ["_weaponItems", false, [true]], ["_backpack", true, [true]], ["_vest", true, [true]], ["_uniform", true, [true]], ["_assignedItems", true, [true]]];
+params [
+    ["_unit", objNull, [objNull]],
+    ["_weaponItems", false, [false]],
+    ["_backpack", true, [false]],
+    ["_vest", true, [false]],
+    ["_uniform", true, [false]],
+    ["_assignedItems", true, [false]],
+    ["_magazines", false, [false]]
+];
 
 private _allItems = if (_assignedItems) then {assignedItems _unit} else {[]};
-if (_uniform) then {_allItems append ((getItemCargo (uniformContainer _unit)) select 0)};
-if (_vest) then {_allItems append ((getItemCargo (vestContainer _unit)) select 0)};
-if (_backpack) then {_allItems append ((getItemCargo (backpackContainer _unit)) select 0)};
+
+if (_uniform) then {
+    _allItems append (getItemCargo uniformContainer _unit select 0);
+};
+
+if (_vest) then {
+    _allItems append (getItemCargo vestContainer _unit select 0);
+};
+
+if (_backpack) then {
+    _allItems append (getItemCargo backpackContainer _unit select 0);
+};
 
 if (_weaponItems) then {
-    _allItems append (primaryWeaponItems _unit);
-    _allItems append (secondaryWeaponItems _unit);
-    _allItems append (handgunItems _unit);
-    _allItems append (primaryWeaponMagazine _unit);
-    _allItems append (secondaryWeaponMagazine _unit);
-    _allItems append (handgunMagazine _unit);
     _allItems append [
-        primaryWeapon _unit, secondaryWeapon _unit, handgunWeapon _unit
-     ];
-     _allItems pushBack (_unit call CBA_fnc_binocularMagazine);
+        primaryWeapon _unit,
+        secondaryWeapon _unit,
+        handgunWeapon _unit // binocular covered by assignedItems
+    ];
+
+    _allItems append primaryWeaponItems _unit;
+    _allItems append secondaryWeaponItems _unit;
+    _allItems append handgunItems _unit;
+
+    if (_magazines) then {
+        _allItems append primaryWeaponMagazine _unit;
+        _allItems append secondaryWeaponMagazine _unit;
+        _allItems append handgunMagazine _unit;
+        _allItems pushBack (_unit call CBA_fnc_binocularMagazine);
+    };
+
+    _allItems = _allItems - [""];
```

Comment:
Might be cheaper to do it at the end, after duplicate removal.
because `-` iterates over everything and copies everything.

Reply:
Or if we know where it comes from (I guess primary/secondary/handgunWeapon) we can do it right there.
Would be cheaper to do in L58 for just the 3 things, than do it for everything.

Reply:
It comes from `p/s/hWeapon` as well as `p/s/hWeaponItems`. The magazine commands report empty arrays if no mag is loaded. I put the array subtraction below the last line that may add empty strings.

The placement is so that the array subtraction command is only executed when the `_weaponItems` flag is set, as this is the only place where the commands may add empty strings. Note that while `p/s/hWeaponItems` has the items in order muzzle/pointer/optics/bipod and fills the gaps with empty strings, assignedItems, while having an order, just changes in array size (and indices) if a specific slot is unoccupied.

If I were to add it to the last line after the `arrayIntersect` command instead, I would slow down the function for every use even when this flag is set to false. I know that a smaller array would speed up the function slightly, but this would mean adding an additional binary command (`-`) for all uses for a small improvement in C++ that itself wouldn't even be a problem in all other use cases. The priority is clear to me.

The only potential optimization here would be moving the weapon(items/magazines) block above the uniform/vest/backpack block.

Reply:
Another potential optimization approach would be creating the initial array as `[""]` and then using `pushBackUnique` everywhere. And then using `deleteAt 0` at the end. However, this would need `appendUnique`, because one source of empty strings are arrays here.

Reply:
Maybe instead store _weaponItems seperately, then remove "" from it.
And only then append it to _allItems. That way you don't have to iterate over uniform/vest/backpack items to remove ""

Reply:
Make a new array here
https://github.com/CBATeam/CBA_A3/pull/1157/files#diff-5281788de7c6251ec13f252782a5e72fR54
And then `_allItems append _newArray` after removing the ""

Reply:
That would mean more commands in total. I don't think there will be any case where this actually would be faster.

Reply:
but only in the weaponsItems case
which I'd say is rare.